feat: add activation checkpointing to unet

ferreirafabio80 opened the pull request:
### Description
Introduces an optional `use_checkpointing` flag in the `UNet` implementation. When enabled, intermediate activations in the encoder–decoder blocks are recomputed during the backward pass instead of being stored in memory. 
 - Implemented via a lightweight `_ActivationCheckpointWrapper` wrapper around sub-blocks.
 - Checkpointing is only applied during training to avoid overhead at inference.

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

Commit:
feat: add optional gradient checkpointing to unet

Commit:
fix: small ruff issue

ferreirafabio80 committed:
Update monai/networks/nets/unet.py

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: Fábio S. Ferreira <ferreira.fabio80@gmail.com>

Commit:
docs: update docstrings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
fix: avoid BatchNorm subblocks

Commit:
fix: revert batch norm changes

ericspod commented on monai/networks/nets/unet.py:
```diff
@@ -24,6 +26,22 @@
+                return cast(torch.Tensor, checkpoint(self.module, x, use_reentrant=False))
+            except TypeError:
+                # Fallback for older PyTorch without `use_reentrant`
+                return cast(torch.Tensor, checkpoint(self.module, x))
```
We don't support versions of PyTorch so old that `use_reentrant` is missing so we can omit this along with the try block.

ferreirafabio80 replied:
Makes sense, thanks.

ericspod commented:
Hi @ferreirafabio80 thanks for the contribution but I would suggest this isn't necessarily the way to go with adapting this class. Perhaps instead you can create a subclass of `UNet` and override the method:

```python
class CheckpointUNet(UNet):
    def _get_connection_block(self, down_path: nn.Module, up_path: nn.Module, subblock: nn.Module) -> nn.Module:
        subblock = _ActivationCheckpointWrapper(subblock)
        return super()._get_connection_block(down_path, up_path, subblock)
```

This would suffice for your own use if you just wanted such a definition. I think the `_ActivationCheckpointWrapper` class may be a good thing to add to the `blocks` submodule instead, so it should have a public name like `CheckpointWrapper`. 

I see also that `checkpoint` is used elsewhere in MONAI already like [here](https://github.com/Project-MONAI/MONAI/blob/53382d838b951069de60e3c62cd7701cbe756aff/monai/networks/nets/autoencoderkl.py#L602) without the checks for training and gradient that you have in your class, so I wonder if these are needed at all?

Commit:
refactor: creates a subclass of UNet and overrides the get connection block method

ferreirafabio80 commented:
Hi @ericspod, thank you for your comments. 

Yes, that also works. I've defined a subclass and overridden the method as you suggested. 

Regarding the `_ActivationCheckpointWrapper` class, should I create a new script in the `blocks` submodule or add it to an existent one?

I was probably being extremely careful with the checks in `_ActivationCheckpointWrapper`, but agree we can drop the checks.

Commit:
chore: remove use checkpointing from doc string

Commit:
fix: linting issues

Commit:
feat: add activation checkpointing to down and up paths to be more efficient

ericspod commented:
> Regarding the `_ActivationCheckpointWrapper` class, should I create a new script in the `blocks` submodule or add it to an existent one?

Sorry for the delay, I think we should put this into a new file in the `monai/networks/blocks` directory since it doesn't really go anywhere else. You can then give it a good name and docstring comments, plus whatever else Coderabbit has said that's reasonable. Thanks!

Commit:
refactor: move activation checkpointing wrapper to blocks

Commit:
chore: add docstrings to checkpointed unet

Commit:
test: add checkpoint unet test

ferreirafabio80 commented:
@ericspod I've moved the wrapper to a different script, added docstrings and a test (which is literally a copy of the unet one). Let me know if this is sensible.

Commit:
fix: change test name

Commit:
fix: simplify test and make sure that checkpoint unet runs well in training

Commit:
fix: set seed

ericspod commented on monai/networks/blocks/activation_checkpointing.py:
```diff
@@ -0,0 +1,41 @@
+        super().__init__()
+        self.module = module
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass with optional activation checkpointing.
+
+        Args:
+            x: Input tensor.
+
+        Returns:
+            Output tensor from the wrapped module.
+        """
+        return cast(torch.Tensor, checkpoint(self.module, x, use_reentrant=False))
```
I think just change the docstring here instead of the code change.

ericspod commented on tests/networks/nets/test_checkpointunet.py:
```diff
@@ -0,0 +1,145 @@
+
+        # checkpointing should reduce activation memory use; we can't directly assert memory savings
+        # but we can confirm no runtime errors and gradients propagate correctly
+        self.assertIsNotNone(grad_norm)
```
This check will never be reached when `grad_norm` is None because the previous one would raise an exception when evaluating `grad_norm.item()`.

ferreirafabio80 replied:
yes, you are right - just removed it.

ericspod commented:
> @ericspod I've moved the wrapper to a different script, added docstrings and a test (which is literally a copy of the unet one). Let me know if this is sensible.

This looks much better, thanks. Please do work on the testing issues and the DCO issue can be left until last.

Commit:
fix: fix testing bugs

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ferreirafabio80 commented:
Thanks. I have fixed the testing issues. What is the easiest way to fix the DCO issue?

KumoLiu commented:
> Thanks. I have fixed the testing issues. What is the easiest way to fix the DCO issue?

Hi @ferreirafabio80, you can refer to the guide here: https://github.com/Project-MONAI/MONAI/pull/8554/checks?check_run_id=54816594479

Commit:
chore: add test docstrings

Signed-off-by: Fabio Ferreira <f.ferreira@qureight.com>

Commit:
DCO Remediation Commit for Fabio Ferreira <f.ferreira@qureight.com>

I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: de2b6bd65ae84c76781ca387f834fc5abb5fb08a
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 66edcb508243f53c4f10af93d6ebfca9a32fe4ef
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: feefcaa3944f56fba163475cc5ef4d0da28ceddf
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: f673ca1453020bba8d9690c3745bb2dc917a806a
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 69540ffe7d16fa81bb30cd0c1c09186c0b59d9da
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 42ec757a76dc9c476b7dd302fb9352eee168b9b3
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: a2e8474abf79552cb4c041c69583261ad16c7049
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 4c4782e6a4d9156f3eeebf90543b2f1699ab3d72
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 515c659ee6f0587d25935ae728195266cf340422
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: da5a3a457c5f55f759778663628ca99044ea93a2
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 43dec884a8615eda52d21fd1070ea2b8db7a8f92
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 84c0f48d0282413731e066850603527977cd622d
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 58055152d92ee503cf42d12691cd9ffa354d9996
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 1aa8e3c3ab4554a578974fa1216b0157d2aba98f
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: 447d9f275e7a02788b1f071a1fefc2aa290b7ad0
I, Fabio Ferreira <f.ferreira@qureight.com>, hereby add my Signed-off-by to this commit: b20a19ec086ae7738d048c2e1a5a95e9e6fdbc43

Signed-off-by: Fabio Ferreira <f.ferreira@qureight.com>

ferreirafabio80 commented:
thank you @KumoLiu. Just fixed it.

ferreirafabio80 commented on tests/networks/nets/test_checkpointunet.py:
```diff
@@ -0,0 +1,187 @@
+            spatial_dims=2, in_channels=1, out_channels=3, channels=(16, 32, 64), strides=(2, 2), num_res_units=0
+        )
+        test_data = torch.randn(16, 1, 32, 32)
+        test_script_save(net, test_data)
```
@ericspod this won't work with the CheckpointUNet because I think torch.utils.checkpoint is not TorchScript-compatible. I can drop the test though.

ericspod replied:
> @ericspod this won't work with the CheckpointUNet because I think torch.utils.checkpoint is not TorchScript-compatible. I can drop the test though.

What does the error look like when you try running this test with the changes coderabbit proposes? It's something that can be fixed possibly, but if not then yes we drop the test and live with the class not being compatible for now at least with Torchscript.

ferreirafabio80 replied:
Here is the error. Not sure how to fix this to be honest.

```concrete_type = <torch.ConcreteModuleType object at 0x322a93ef0>
method_stubs = [ScriptMethodStub(resolution_callback=<function createResolutionCallbackFromEnv.<locals>.<lambda> at 0x322b4b940>, def..., track_running_stats=False)
      (D): Dropout(p=0.0, inplace=False)
      (A): PReLU(num_parameters=1)
    )
  )
)>)]
property_stubs = []

    def create_methods_and_properties_from_stubs(
        concrete_type, method_stubs, property_stubs
    ):
        method_defs = [m.def_ for m in method_stubs]
        method_rcbs = [m.resolution_callback for m in method_stubs]
        method_defaults = [get_default_args(m.original_method) for m in method_stubs]
    
        property_defs = [p.def_ for p in property_stubs]
        property_rcbs = [p.resolution_callback for p in property_stubs]
    
>       concrete_type._create_methods_and_properties(
            property_defs, property_rcbs, method_defs, method_rcbs, method_defaults
        )
E       RuntimeError: 
E       builtin cannot be used as a value:
E         File "/Users/FabioFerreria/Projects/MONAI/monai/networks/blocks/activation_checkpointing.py", line 41
E                   Output tensor from the wrapped module.
E               """
E               return cast(torch.Tensor, checkpoint(self.module, x, use_reentrant=False))
E                           ~~~~~~~~~~~~ <--- HERE

ericspod replied:
Ok let's remove the Torchscript test, it's less relevant now that Torchscript is being deprecated.

ferreirafabio80 replied:
cool, done.

Commit:
fix: remove test script save

Signed-off-by: Fabio Ferreira <f.ferreira@qureight.com>

Commit:
fix: tighten tolerance for numerical equivalence

Signed-off-by: Fabio Ferreira <f.ferreira@qureight.com>

ericspod commented:
Hi @ferreirafabio80 it's looking pretty good here, if you can resolve the last conversation items with coderabbit (eg. update docstrings) and others we can merge this soon I think.

Commit:
chore: update doc strings

Signed-off-by: Fabio Ferreira <f.ferreira@qureight.com>

ferreirafabio80 commented:
@ericspod sorry, missed that one. All addressed now.

KumoLiu committed:
Merge branch 'dev' into feat/add_activation_checkpointing_to_unet

KumoLiu commented:
/build
